packages/protocol-parser/src/Schema.ts

```diff
@@ -0,0 +1,160 @@
+  }
+
+  isEmpty(): boolean {
+    return this.staticFields.length === 0 && this.dynamicFields.length === 0;
+  }
+
+  toHex(): Hex {
+    const staticSchemaTypes = this.staticFields.map((abiType) => schemaAbiTypes.indexOf(abiType));
+    const dynamicSchemaTypes = this.dynamicFields.map((abiType) => schemaAbiTypes.indexOf(abiType));
+    return `0x${[
+      this.staticDataLength().toString(16).padStart(4, "0"),
+      this.staticFields.length.toString(16).padStart(2, "0"),
+      this.dynamicFields.length.toString(16).padStart(2, "0"),
+      ...staticSchemaTypes.map((schemaType) => schemaType.toString(16).padStart(2, "0")),
+      ...dynamicSchemaTypes.map((schemaType) => schemaType.toString(16).padStart(2, "0")),
+    ]
+      .join("")
+      .padEnd(64, "0")}`;
+  }
+
+  decodeRecord(data: Hex): readonly (StaticPrimitiveType | DynamicPrimitiveType)[] {
+    const values: (StaticPrimitiveType | DynamicPrimitiveType)[] = [];
+
+    let bytesOffset = 0;
+    this.staticFields.forEach((fieldType) => {
+      const fieldByteLength = staticAbiTypeToByteLength[fieldType];
+      const value = decodeStaticField(fieldType, sliceHex(data, bytesOffset, bytesOffset + fieldByteLength));
+      bytesOffset += fieldByteLength;
+      values.push(value);
+    });
+
+    // Warn user if static data length doesn't match the schema, because data corruption might be possible.
+    const actualStaticDataLength = bytesOffset;
+    if (actualStaticDataLength !== this.staticDataLength()) {
+      console.warn(
+        "Decoded static data length does not match schema's expected static data length. Data may get corrupted. Is `getStaticByteLength` outdated?",
+        {
+          expectedLength: this.staticDataLength,
+          actualLength: actualStaticDataLength,
+          bytesOffset,
+        }
+      );
+    }
+
+    if (this.dynamicFields.length > 0) {
+      const dataLayout = hexToPackedCounter(sliceHex(data, bytesOffset, bytesOffset + 32));
+      bytesOffset += 32;
+
+      this.dynamicFields.forEach((fieldType, i) => {
+        const dataLength = dataLayout.fieldByteLengths[i];
+        const value = decodeDynamicField(fieldType, sliceHex(data, bytesOffset, bytesOffset + dataLength));
+        bytesOffset += dataLength;
+        values.push(value);
+      });
+
+      // Warn user if dynamic data length doesn't match the schema, because data corruption might be possible.
+      const actualDynamicDataLength = bytesOffset - 32 - actualStaticDataLength;
+      // TODO: refactor this so we don't break for bytes offsets >UINT40
+      if (BigInt(actualDynamicDataLength) !== dataLayout.totalByteLength) {
+        console.warn(
+          "Decoded dynamic data length does not match data layout's expected data length. Data may get corrupted. Did the data layout change?",
+          {
+            expectedLength: dataLayout.totalByteLength,
+            actualLength: actualDynamicDataLength,
+            bytesOffset,
+          }
+        );
+      }
```

Comment:
I think since this is new code, I can probably do that now. This was copied over from another area of the codebase, where it was a much bigger refactor.

Reply:
ahh I remember now: viem's `sliceHex` operations expect a `number`, so that's why `bytesOffset` is a number

moving it to `bigint` means that I can't use `sliceHex` unless downcasting, which has the same potential issue.

Reply:
`uint40` max is `1099511627775`, which means we'd need a 1TB string to hit this issue. Feels like we could punt on this for now.